test/raft/distributed_test.cc

```diff
@@ -31,5 +40,142 @@ requires (typename M::state_t s, typename M::input_t i) {
+            } else {
+                // This is not the leader on which the command was submitted,
+                // or it is but the client already gave up on us and deallocated the channel.
+                // In any case we simply drop the output.
+            }
+
+            if (_aborted) {
+                co_return;
+            }
+
+            co_await make_ready_future<>(); // maybe yield
+        }
+    }
+
+    future<raft::snapshot_id> take_snapshot() override {
+        auto id = raft::snapshot_id{utils::make_random_uuid()};
+        _snapshots[id] = _val;
+        co_return id;
+    }
+
+    void drop_snapshot(raft::snapshot_id id) override {
+        _snapshots.erase(id);
+    }
+
+    future<> load_snapshot(raft::snapshot_id id) override {
+        auto it = _snapshots.find(id);
+        assert(it != _snapshots.end()); // dunno if the snapshot can actually be missing
+        _val = it->second;
+        co_return;
+    }
+
+    future<> abort() override {
+        _aborted = true;
+        // TODO: the semantics of abort are not clear.
+        // If apply is in progress, should we wait for it to finish here? (it will finish on the next iteration over cmds)
+        co_return;
+    }
+
+    template <typename F>
+    future<typename M::output_t> with_output_channel(F&& f) {
+        promise<typename M::output_t> p;
+        auto fut = p.get_future();
+        auto cmd_id = utils::make_random_uuid();
+        auto [_, inserted] = _output_channels.emplace(cmd_id, std::move(p));
+        assert(inserted);
+
+        return std::forward<F>(f)(cmd_id, std::move(fut)).finally([cmd_id, this] () {
+            // TODO ensure `this` doesn't get destroyed in the meantime (can it)?
+            _output_channels.erase(cmd_id);
+        });
+    }
+};
+
+// TODO: serializable concept?
+template <PureStateMachine M>
+raft::command make_command(const cmd_id_t& cmd_id, const typename M::input_t& input) {
+    raft::command cmd;
+    ser::serialize(cmd, cmd_id);
+    ser::serialize(cmd, input);
+    return cmd;
+}
+
+// TODO: handle other errors?
+template <PureStateMachine M>
+using call_result_t = std::variant<typename M::output_t, timed_out_error, raft::not_a_leader>;
+
+template <PureStateMachine M>
+future<call_result_t<M>> call(
+        typename M::input_t input,
+        raft::clock_type::time_point timeout,
+        raft::server& server,
+        impure_state_machine<M>& sm) {
+
+    return sm.with_output_channel([input = std::move(input), timeout, &server] (cmd_id_t cmd_id, future<typename M::output_t> f) {
+        return with_timeout(timeout, [input = std::move(input), &server, cmd_id] (future<typename M::output_t> f) -> future<typename M::output_t> {
+            co_await server.add_entry(
+                    make_command<M>(cmd_id, std::move(input)),
+                    raft::wait_type::applied);
+
+            co_return co_await std::move(f);
+        }(std::move(f)));
+    }).then([] (typename M::output_t o) {
```

Comment:
Here as well? Why use `then()` and `handle_exception()` instead of simple try{}catch() + co-routines?